packages/SwingSet/src/kernel/collectionManager.js

```diff
@@ -0,0 +1,891 @@
+/* global BigUint64Array */
+// @ts-check
+import { assert, details as X, q } from '@agoric/assert';
+import {
+  getRankCover,
+  assertKeyPattern,
+  assertPattern,
+  matches,
+  compareRank,
+  M,
+} from '@agoric/store';
+import {
+  Far,
+  passStyleOf,
+  nameForPassableSymbol,
+  passableSymbolForName,
+} from '@agoric/marshal';
+import { parseVatSlot } from '../parseVatSlots.js';
+
+function zeroPad(n, size) {
+  const str = `000000000000000000${n}`;
+  return str.substring(str.length - size);
+}
+
+// This is the JavaScript analog to a C union: a way to map between a float as a
+// number and the bits that represent the float as a buffer full of bytes.  Note
+// that the mutation of static state here makes this invalid Jessie code, but
+// doing it this way saves the nugatory and gratuitous allocations that would
+// happen every time you do a conversion -- and in practical terms it's safe
+// because we put the value in one side and then immediately take it out the
+// other; there is no actual state retained in the classic sense and thus no
+// re-entrancy issue.
+const asNumber = new Float64Array(1);
+const asBits = new BigUint64Array(asNumber.buffer);
+
+function numberToDBEntryKey(n) {
+  asNumber[0] = n;
+  let bits = asBits[0];
+  if (n < 0) {
+    // XXX Why is the no-bitwise lint rule even a thing??
+    // eslint-disable-next-line no-bitwise
+    bits ^= 0xffffffffffffffffn;
+  } else {
+    // eslint-disable-next-line no-bitwise
+    bits ^= 0x8000000000000000n;
+  }
+  return `f${zeroPad(bits.toString(16), 16)}`;
+}
+
+function dbEntryKeyToNumber(k) {
+  let bits = BigInt(`0x${k.substring(1)}`);
+  if (k[0] < '8') {
+    // eslint-disable-next-line no-bitwise
+    bits ^= 0xffffffffffffffffn;
+  } else {
+    // eslint-disable-next-line no-bitwise
+    bits ^= 0x8000000000000000n;
+  }
+  asBits[0] = bits;
+  return asNumber[0];
+}
+
+const BIGINT_TAG_LEN = 10;
+const BIGINT_LEN_MODULUS = 10 ** BIGINT_TAG_LEN;
+
+function bigintToDBEntryKey(n) {
```

Comment:
Yeah, it probably *should* be written up, and I will do so.  However, I'm thinking that since this is implementation description, the place to write it up is in the implementation itself.  If it turns out we want to have a separate document we can extract the comment text for such a purpose later.